### ISSUE TITLE
CF Metadata 

### ISSUE THREAD

**Issue by aph42** [pygeode/pygeode#80]
At present the metadata encoding and decoding happens all in a bit messy set of functions in cfmeta.py. At least part of the logic should be handled by object-oriented stuff within axis classes. Part of this is just organizational - as much of the logic that is specific to, say, organizing time axes, should be within the timeaxis.py set of objects. But I think this would make a couple things easier.

For one it should be possible to allow custom/new axis objects to be able to handle CF metadata encoding and decoding without having to open up the cfmeta.py central logic (cf. #68).  

Moreover I think this metadata whenever possible should remain attached and accessible to the axis objects in Pygeode (rather than consumed and recreated on read in and write out). If this logic is in the axis objects, every new Pres axis, for instance, can have relevant cf metadata produced on its construction. This might also be a useful place to use the configuration file ( #62 ) if there are defaults that need to be chosen. 

It'd be nice to have a library of CF stuff to attach to variables too, so that things like standard_names and units and stuff are more easily populated, though how to do this is less obvious (maybe having some kind of standard dictionary/configuration file to populate attributes easily?).

**Comment by neishm** [comment]
I agree that the logic for encoding axes should be put in the axes themselves.  For decoding, perhaps we'd also need some kind of lookup table of standard names, pointing to the particular Axis that can handle the data?

I think the original reason why the CF metadata was stripped out, and why the encoding/decoding was in a separate `cfmeta` module, was to keep netCDF-specific stuff out of the PyGeode objects.  We could, in theory, have a different set of "standard" attributes for other file formats, or even support [multiple netCDF conventions](http://www.unidata.ucar.edu/software/netcdf/conventions.html) through different encoding/decoding modules.  In practice, though, we mostly deal with netCDF files, and CF is the recommended standard to follow.

Variables, on the other hand, would be very difficult to handle.  There is a table of standard names in the CF convention, but I hardly ever see them used in practice.  A short-term solution could be to allow the user to define a custom dictionary for their data source(s), so they can map the local name to the CF standard name (and other attributes).  However, these would be entirely user-defined.  There would literally be hundreds of such tables, and there'd be little point trying to capture them all in our package.  Off the top of my head, I've dealt with 11 different data sources within the last month, each with their own naming convention for variables.  The axes (usually) get recognized automatically, though.
